File: go/frontend/graphic.go
package frontend

import (
	"github.com/hajimehoshi/ebiten/v2"

	"github.com/ignatiusreza/chip8/go/chip8"
)

const scale = 10

// Graphic is the emulator window, showing the 64x32 display scaled 10x.
type Graphic struct {
	image  *ebiten.Image
	pixels []byte
}

func NewGraphic(title string) *Graphic {
	ebiten.SetWindowTitle(title)
	ebiten.SetWindowSize(chip8.Width*scale, chip8.Height*scale)
	return &Graphic{
		image:  ebiten.NewImage(chip8.Width, chip8.Height),
		pixels: make([]byte, chip8.BuffLength*4),
	}
}

// Draw redraws the display into the window, re-reading the frame buffer only if it changed.
func (g *Graphic) Draw(screen *ebiten.Image, display *chip8.Display) {
	if display.TakeInvalidated() {
		for y := range chip8.Height {
			for x := range chip8.Width {
				var c byte
				if display.Get(x, y) {
					c = 0xFF
				}
				i := (x + y*chip8.Width) * 4
				g.pixels[i], g.pixels[i+1], g.pixels[i+2], g.pixels[i+3] = c, c, c, 0xFF
			}
		}
		g.image.WritePixels(g.pixels)
	}

	op := &ebiten.DrawImageOptions{}
	op.GeoM.Scale(scale, scale)
	screen.DrawImage(g.image, op)
}

// Layout returns the window's logical size.
func (g *Graphic) Layout() (int, int) {
	return chip8.Width * scale, chip8.Height * scale
}

File: rust/src/frontend/graphic.rs
use chip8::Display;
use minifb::{Scale, Window, WindowOptions};

const FPS: usize = 60;
const BLACK: u32 = 0x000000;
const WHITE: u32 = 0xFFFFFF;

/// The emulator window, showing the 64x32 display scaled 8x.
pub struct Graphic {
    window: Window,
    buffer: Vec<u32>,
}

impl Graphic {
    pub fn new(title: &str) -> Result<Self, minifb::Error> {
        let mut window = Window::new(
            title,
            Display::WIDTH,
            Display::HEIGHT,
            WindowOptions {
                scale: Scale::X8,
                ..WindowOptions::default()
            },
        )?;
        window.set_target_fps(FPS);
        Ok(Graphic {
            window,
            buffer: vec![BLACK; Display::BUFF_LENGTH],
        })
    }

    /// The window, which also receives the keyboard input.
    pub fn window(&self) -> &Window {
        &self.window
    }

    pub fn is_open(&self) -> bool {
        self.window.is_open()
    }

    /// Redraws the display if it changed. Either way the window is updated, which
    /// polls input and keeps the loop at 60fps.
    pub fn update(&mut self, display: &mut Display) -> Result<(), minifb::Error> {
        if !display.take_invalidated() {
            self.window.update();
            return Ok(());
        }

        for y in 0..Display::HEIGHT {
            for x in 0..Display::WIDTH {
                self.buffer[x + y * Display::WIDTH] = if display.get(x, y) { WHITE } else { BLACK };
            }
        }
        self.window
            .update_with_buffer(&self.buffer, Display::WIDTH, Display::HEIGHT)
    }
}
